pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go

```diff
@@ -83,10 +85,50 @@ func (s *preFilterState) Clone() framework.StateData {
 // failed by this plugin schedulable.
 func (pl *NodeAffinity) EventsToRegister() []framework.ClusterEventWithHint {
 	return []framework.ClusterEventWithHint{
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
 	}
 }
 
+// isSchedulableAfterNodeChange is invoked whenever a node changed. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *NodeAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterNodeChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	if pl.addedNodeSelector != nil && !pl.addedNodeSelector.Match(modifiedNode) {
+		logger.V(4).Info(errReasonEnforced, "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		return framework.QueueSkip
+	}
+
+	requiredNodeAffinity := nodeaffinity.GetRequiredNodeAffinity(pod)
+	isMatched, err := requiredNodeAffinity.Match(modifiedNode)
+	if err != nil {
+		logger.Error(err, "failed to match pod to the modified node", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		return framework.QueueAfterBackoff
+	}
```

Comment:
if isMatched == true, we can soon return Skip, no need to check `wasMatched`

Reply:
@sanposhiho if isMatched == true, why not return framework.QueueAfterBackoff?

Reply:
and if we return here, `if !wasMatched && isMatched ` in line 122 will never pass

Reply:
> if isMatched == true, why not return framework.QueueAfterBackoff?

Right 😅 I meant to say that

Reply:
Wait, No, sorry I caused confusion here.

Let's look at your past version again:
https://github.com/kubernetes/kubernetes/commit/4fc4ef3e53f8b4e7c52c6ab8a6464df3484f1909

What I really wanted to say was, if isMatched == ~~true~~ **false**, you can soon return Skip. 

1. check `isMatched`
2. **if isMatched == false, return Skip early.** (my suggestion)
3. check `wasMatched`
4. if !wasMatched && isMatched, return QueueAfterBackoff, otherwise return Skip.

Reply:
@sanposhiho I got you and updated it.

Reply:
@sanposhiho just to clarify: we need to check `wasMatched` is b/c unschedulePlugins only tell us which plugins it failed by, but it doesn't tell us which **node(s)** the pod failed on NodeAffinity plugin. So supposed the pod failed on node1 due to NodeAffinity, while here the node event is on nodeX, so wasMatched should be examined.